runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/logging/GrpcLoggingService.java

```diff
@@ -0,0 +1,120 @@
+ */
+
+package org.apache.beam.runners.fnexecution.logging;
+
+import com.google.common.collect.ImmutableSet;
+import io.grpc.stub.StreamObserver;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.LogControl;
+import org.apache.beam.model.fnexecution.v1.BeamFnLoggingGrpc;
+import org.apache.beam.runners.fnexecution.FnService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** An implementation of the Beam Fn Logging Service over GRPC. */
+public class GrpcLoggingService extends BeamFnLoggingGrpc.BeamFnLoggingImplBase
+    implements FnService {
+  private static final Logger LOGGER = LoggerFactory.getLogger(GrpcLoggingService.class);
+
+  public static GrpcLoggingService forWriter(LogWriter writer) {
+    return new GrpcLoggingService(writer);
+  }
+
+  private final LogWriter clientLogger;
+  private final ConcurrentMap<InboundObserver, StreamObserver<LogControl>> connectedClients;
+
+  private GrpcLoggingService(LogWriter clientLogger) {
+    this.clientLogger = clientLogger;
+    connectedClients = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void close() throws Exception {
+    Set<InboundObserver> remainingClients = ImmutableSet.copyOf(connectedClients.keySet());
+    if (!remainingClients.isEmpty()) {
+      LOGGER.info(
+          "{} Beam Fn Logging clients still connected during shutdown.", remainingClients.size());
+
+      // Signal server shutting down to all remaining connected clients.
+      for (InboundObserver client : remainingClients) {
+        // We remove these from the connected clients map to prevent a race between
+        // this close method and the InboundObserver calling a terminal method on the
+        // StreamObserver. If we removed it, then we are responsible for the terminal call.
+        completeIfNotNull(connectedClients.remove(client));
+      }
+    }
+  }
+
+  @Override
+  public StreamObserver<BeamFnApi.LogEntry.List> logging(
+      StreamObserver<BeamFnApi.LogControl> outboundObserver) {
+    LOGGER.info("Beam Fn Logging client connected.");
+    InboundObserver inboundObserver = new InboundObserver();
+    connectedClients.put(inboundObserver, outboundObserver);
+    return inboundObserver;
+  }
+
+  private void completeIfNotNull(StreamObserver<BeamFnApi.LogControl> outboundObserver) {
+    if (outboundObserver != null) {
+      try {
+        outboundObserver.onCompleted();
+      } catch (RuntimeException ignored) {
+        // Completing outbound observer failed, ignoring failure and continuing
+        LOGGER.warn("Beam Fn Logging client failed to be complete.", ignored);
+      }
+    }
+  }
+
+  /**
+   * An inbound {@link StreamObserver} that forwards incoming messages to the client logger.
+   *
+   * <p>Maintains the set of connected clients. Mutually hangs up on clients that have errored
```

Comment:
This class doesn't maintain the set of connected clients.

Reply:
Rm'd that part